File: matlab/lib/kpctoolbox/dmap/dmap_moment.m
function MOMENTS=dmap_moment(DMAP,ORDERS)
% MOMENTS=dmap_moment(DMAP,ORDERS) - Compute raw moments of
% inter-arrival times for a discrete-time MAP
%
%  Input:
%  DMAP: a D-MAP in the form of {D0,D1}
%  ORDERS: set of moment orders (1=>E[X], 2=>E[X^2], 3=>E[X^3])
%
%  Output:
%  MOMENTS: moments returned in the same order of ORDERS
%
%  NOTE: the previous formula factorial(i)*al*(I-D0)^-i*e returned neither the
%  raw nor the falling-factorial moment for order>=2 (e.g. it gave 4.447 where
%  the true E[X^2] is 2.958). The raw moments below match the JAR
%  implementation and a brute-force sum over P(X=k)=al*D0^(k-1)*D1*e.

D0=DMAP{1};
D1=DMAP{2};
N=size(D0,1);
e=ones(N,1);
P=inv(eye(N)-D0)*D1;
al=dtmc_solve(P);
A=inv(eye(N)-D0);
Ae=A*e;
m1=al*Ae;
for t=1:length(ORDERS)
    i=ORDERS(t);
    if any(isnan(D0(:)))
        MOMENTS(t)=NaN;
    else
        switch i
            case 1
                MOMENTS(t)=m1;
            case 2
                MOMENTS(t)=2*al*A*Ae - m1;
            case 3
                MOMENTS(t)=6*al*A*A*Ae - 6*al*A*Ae + m1;
            otherwise
                error('dmap_moment: raw moments of order > 3 not implemented');
        end
    end
end
end

File: jar/src/main/java/jline/api/mam/Dmap_moment.java
/**
 * Moment computation for discrete-time MAPs.
 *
 * @since LINE 3.0
 */
package jline.api.mam;

import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class Dmap_moment {
    private Dmap_moment() {}

    /**
     * Computes the k-th raw moment of the inter-arrival time of a discrete MAP.
     *
     *   E[T]   = pi * (I-D0)^{-1} * e
     *   E[T^2] = 2*pi*(I-D0)^{-2}*e - E[T]
     *   E[T^3] = 6*pi*(I-D0)^{-3}*e - 6*pi*(I-D0)^{-2}*e + E[T]
     */
    public static double dmap_moment(Matrix D0, Matrix D1, int order) {
        int n = D0.getNumRows();
        Matrix I = Matrix.eye(n);
        Matrix ImD0inv = I.add(-1.0, D0).inv();
        Matrix e = Matrix.ones(n, 1);
        Matrix pi = Dmap_pie.dmap_pie(D0, D1);

        switch (order) {
            case 1:
                return pi.mult(ImD0inv).mult(e).toDouble();
            case 2: {
                double m1 = pi.mult(ImD0inv).mult(e).toDouble();
                return 2.0 * pi.mult(ImD0inv.mult(ImD0inv)).mult(e).toDouble() - m1;
            }
            case 3: {
                double m1 = pi.mult(ImD0inv).mult(e).toDouble();
                Matrix ImD0inv2 = ImD0inv.mult(ImD0inv);
                return 6.0 * pi.mult(ImD0inv2.mult(ImD0inv)).mult(e).toDouble()
                        - 6.0 * pi.mult(ImD0inv2).mult(e).toDouble() + m1;
            }
            default:
                throw new IllegalArgumentException("Moments of order > 3 not implemented for DMAP");
        }
    }

    public static double dmap_moment(MatrixCell DMAP, int order) {
        return dmap_moment(DMAP.get(0), DMAP.get(1), order);
    }
}
